src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs

```diff
@@ -447,101 +447,219 @@ public void Remove([NotNull] Operation operation, [NotNull] object objectToApply
+            // get path result
+            var pathResult = GetActualPropertyPath(
+                path,
+                objectToApplyTo,
+                operationToReport);
 
-            if (path.EndsWith("/-"))
+            if (pathResult == null)
             {
-                removeFromList = true;
-                actualPathToProperty = path.Substring(0, path.Length - 2);
+                return null;
             }
-            else
-            {
-                positionAsInteger = GetNumericEnd(path);
 
-                if (positionAsInteger > -1)
-                {
-                    actualPathToProperty = path.Substring(0,
-                        path.IndexOf('/' + positionAsInteger.ToString()));
-                }
-            }
+            var removeFromList = pathResult.ExecuteAtEnd;
+            var positionAsInteger = pathResult.NumericEnd;
+            var actualPathToProperty = pathResult.PathToProperty;
 
-            var patchProperty = FindPropertyAndParent(objectToApplyTo, actualPathToProperty);
+            var treeAnalysisResult = new ObjectTreeAnalysisResult(
+               objectToApplyTo,
+               actualPathToProperty,
+               ContractResolver);
 
-            // does the target location exist?
-            if (!CheckIfPropertyExists(patchProperty, objectToApplyTo, operationToReport, path))
+            if (!treeAnalysisResult.IsValidPathForRemove)
             {
-                return;
+                LogError(new JsonPatchError(
+                    objectToApplyTo,
+                    operationToReport,
+                    Resources.FormatPropertyCannotBeRemoved(path)));
+                return null;
             }
 
-            // get the property, and remove it - in this case, for DTO's, that means setting
-            // it to null or its default value; in case of an array, remove at provided index
-            // or at the end.
-            if (removeFromList || positionAsInteger > -1)
+            if (treeAnalysisResult.UseDynamicLogic)
             {
-                // what if it's an array but there's no position??
-                if (IsNonStringArray(patchProperty.Property.PropertyType))
+                // if it's not an array, we can remove the property from
+                // the dictionary.  If it's an array, we need to check the position first.
+                if (removeFromList || positionAsInteger > -1)
                 {
-                    // now, get the generic type of the IList<> from Property type.
-                    var genericTypeOfArray = GetIListType(patchProperty.Property.PropertyType);
+                    var propertyValue = treeAnalysisResult.Container
+                        .GetValueForCaseInsensitiveKey(treeAnalysisResult.PropertyPathInParent);
+                    var typeOfPathProperty = treeAnalysisResult.Container
+                            .GetValueForCaseInsensitiveKey(treeAnalysisResult.PropertyPathInParent).GetType();
```

Comment:
instead of calling into the dictionary twice just call GetType() on the object. Also what if this returns null?